src/move.cpp

```diff
@@ -898,92 +897,131 @@ AtomicSwapCharge::AtomicSwapCharge(Space &spc, std::string name, std::string cit
-                Change::data d;
-                d.index = Faunus::distance(spc.groups.begin(), it); // integer *index* of moved group
-                d.all = true;                                       // *all* atoms in group were moved
-                change.groups.push_back(d);                         // add to list of moved groups
-            }
+void TranslateRotate::_move(Change &change) {
+    if (auto group = findRandomMolecule()) { // note that group is of type std::optional
+        latest_displacement_squared = translateMolecule(group->get());
+        latest_rotation_angle_squared = rotateMolecule(group->get());
+        if (latest_displacement_squared > 0.0 || latest_rotation_angle_squared > 0.0) { // report changes
+            auto &change_data = change.groups.emplace_back();
+            change_data.index = spc.getGroupIndex(group->get());     // integer *index* of moved group
+            change_data.all = true;                                  // *all* atoms in group were moved
+            change_data.internal = false;                            // internal energy is unchanged
+        }
 #ifndef NDEBUG
-            // check if mass center is correctly moved and can be re-calculated
-            Point cm_recalculated = Geometry::massCenter(it->begin(), it->end(), spc.geo.getBoundaryFunc(), -it->cm);
-            double should_be_small = spc.geo.sqdist(it->cm, cm_recalculated);
-            if (should_be_small > 1e-6) {
-                std::cerr << "cm recalculated: " << cm_recalculated.transpose() << "\n";
-                std::cerr << "cm in group:     " << it->cm.transpose() << "\n";
-                assert(false);
-            }
-//            assert(spc.geo.sqdist(it->cm, Geometry::massCenter(it->begin(), it->end(), spc.geo.getBoundaryFunc(),
-//                                                               -it->cm)) < 1e-6);////////////////
+        sanityCheck(group->get());
 #endif
-        }
+    } else {
+        latest_displacement_squared = 0.0;   // these are used to track mean squared
+        latest_rotation_angle_squared = 0.0; // translational and rotational displacements
+    }
+}
+
+void TranslateRotate::sanityCheck(const Space::Tgroup &group) const {
+    Point cm_recalculated = Geometry::massCenter(group.begin(), group.end(), spc.geo.getBoundaryFunc(), -group.cm);
+    double should_be_small = spc.geo.sqdist(group.cm, cm_recalculated);
+    if (should_be_small > 1e-6) {
+        std::cerr << "cm recalculated: " << cm_recalculated.transpose() << "\n";
+        std::cerr << "cm in group:     " << group.cm.transpose() << "\n";
+        assert(false);
     }
 }
 
+void TranslateRotate::_accept(Change &) {
+    mean_squared_displacement += latest_displacement_squared;
+    mean_squared_rotation_angle += latest_rotation_angle_squared;
+}
+
+void TranslateRotate::_reject(Change &) {
+    mean_squared_displacement += 0.0;
+    mean_squared_rotation_angle += 0.0;
+}
```

Comment:
Matter of interpretation, but shall a rejected move contribute to the mean displacement?

Reply:
Yes, because the old configuration should be counted twice in the MC ensemble. This way the mean squared displacement resembles diffusion.